android/src/main/java/io/grpc/android/AndroidChannelBuilder.java

```diff
@@ -0,0 +1,255 @@
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ConnectivityState;
+import io.grpc.ExperimentalApi;
+import io.grpc.ForwardingChannelBuilder;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.MethodDescriptor;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.okhttp.OkHttpChannelBuilder;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Builds a {@link ManagedChannel} that automatically monitors the Android device's network state.
+ * Network changes are propagated to the underlying OkHttp-backed {@ManagedChannel} to smoothly
+ * handle intermittent network failures.
+ *
+ * <p>gRPC Cronet users should use {@code CronetChannelBuilder} directly, as Cronet itself monitors
+ * the device network state.
+ *
+ * <p>Requires the Android ACCESS_NETWORK_STATE permission.
+ *
+ * @since 1.12.0
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/4056")
+public final class AndroidChannelBuilder extends ForwardingChannelBuilder<AndroidChannelBuilder> {
+
+  private final ManagedChannelBuilder delegateBuilder;
+  private final Context context;
+
+  /** Always fails. Call {@link #forAddress(String, int, Context)} instead. */
+  public static AndroidChannelBuilder forTarget(String target) {
+    throw new UnsupportedOperationException("call forTarget(String, Context) instead");
+  }
+
+  /** Always fails. Call {@link #forAddress(String, int, Context)} instead. */
+  public static AndroidChannelBuilder forAddress(String name, int port) {
+    throw new UnsupportedOperationException("call forAddress(String, int, Context) instead");
+  }
+
+  /** Creates a new builder for the given target and Android context. */
+  public static final AndroidChannelBuilder forTarget(String target, Context context) {
+    return new AndroidChannelBuilder(target, context);
+  }
+
+  /** Creates a new builder for the given host, port, and Android context. */
+  public static AndroidChannelBuilder forAddress(String name, int port, Context context) {
+    return forTarget(GrpcUtil.authorityFromHostAndPort(name, port), context);
+  }
+
+  private AndroidChannelBuilder(String target, Context context) {
+    delegateBuilder = OkHttpChannelBuilder.forTarget(target);
+    this.context = context;
+  }
+
+  @Override
+  protected ManagedChannelBuilder<?> delegate() {
+    return delegateBuilder;
+  }
+
+  @Override
+  public ManagedChannel build() {
+    return new AndroidChannel(delegateBuilder.build(), context);
+  }
+
+  /**
+   * Wraps an OkHttp channel and handles invoking the appropriate methods (e.g., {@link
+   * ManagedChannel#resetConnectBackoff}) when the device network state changes.
+   */
+  @VisibleForTesting
+  static final class AndroidChannel extends ManagedChannel {
+
+    private final ManagedChannel delegate;
+    private final Context context;
+    private final ConnectivityManager connectivityManager;
+
+    private DefaultNetworkCallback defaultNetworkCallback;
```

Comment:
Since these are only used for shut down, you could consider just storing a Runnable for clean-up. It's fine as-is though.

(I'm mainly considering it because of the `Build.VERSION.SDK_INT >= Build.VERSION_CODES.N` check during shut down)

Reply:
Done